File: pkg/lexer/lexer.go
package lexer

import (
	"fmt"
	"unicode"
	"unicode/utf8"
)

// Lexer tokenizes Nex source text.
type Lexer struct {
	input        string
	position     int  // current byte offset (points at ch)
	readPosition int  // next byte to read
	ch           rune // current character
	line         int
	column       int
	errors       []string
}

// New creates a Lexer for the given source input.
func New(input string) *Lexer {
	l := &Lexer{
		input:  input,
		line:   1,
		column: 0,
	}
	l.readChar()
	return l
}

// Errors returns lexing errors encountered so far.
func (l *Lexer) Errors() []string {
	return l.errors
}

// NextToken advances the lexer and returns the next token.
func (l *Lexer) NextToken() Token {
	l.skipWhitespaceAndComments()

	tok := Token{Line: l.line, Column: l.column}

	switch l.ch {
	case '=':
		if l.peekChar() == '=' {
			ch := l.ch
			l.readChar()
			tok.Type = EQ
			tok.Literal = string(ch) + string(l.ch)
		} else {
			tok = l.newToken(Assign, l.ch)
		}
	case '+':
		tok = l.newToken(Plus, l.ch)
	case '-':
		if l.peekChar() == '>' {
			l.readChar()
			tok.Type = Arrow
			tok.Literal = "->"
		} else {
			tok = l.newToken(Minus, l.ch)
		}
	case '*':
		tok = l.newToken(Asterisk, l.ch)
	case '/':
		tok = l.newToken(Slash, l.ch)
	case '%':
		tok = l.newToken(Percent, l.ch)
	case '!':
		if l.peekChar() == '=' {
			ch := l.ch
			l.readChar()
			tok.Type = NotEQ
			tok.Literal = string(ch) + string(l.ch)
		} else {
			tok = l.newToken(Bang, l.ch)
		}
	case '<':
		if l.peekChar() == '=' {
			ch := l.ch
			l.readChar()
			tok.Type = LTE
			tok.Literal = string(ch) + string(l.ch)
		} else {
			tok = l.newToken(LT, l.ch)
		}
	case '>':
		if l.peekChar() == '=' {
			ch := l.ch
			l.readChar()
			tok.Type = GTE
			tok.Literal = string(ch) + string(l.ch)
		} else {
			tok = l.newToken(GT, l.ch)
		}
	case '&':
		if l.peekChar() == '&' {
			ch := l.ch
			l.readChar()
			tok.Type = And
			tok.Literal = string(ch) + string(l.ch)
		} else {
			l.errors = append(l.errors, fmt.Sprintf("illegal character %q at line %d, column %d", l.ch, l.line, l.column))
			tok = l.newToken(Illegal, l.ch)
		}
	case '|':
		if l.peekChar() == '|' {
			ch := l.ch
			l.readChar()
			tok.Type = Or
			tok.Literal = string(ch) + string(l.ch)
		} else if l.peekChar() == '>' {
			l.readChar()
			tok.Type = Pipe
			tok.Literal = "|>"
		} else {
			l.errors = append(l.errors, fmt.Sprintf("illegal character %q at line %d, column %d", l.ch, l.line, l.column))
			tok = l.newToken(Illegal, l.ch)
		}
	case '.':
		tok = l.newToken(Dot, l.ch)
	case '?':
		tok = l.newToken(Question, l.ch)
	case ',':
		tok = l.newToken(Comma, l.ch)
	case ';':
		tok = l.newToken(Semicolon, l.ch)
	case ':':
		tok = l.newToken(Colon, l.ch)
	case '(':
		tok = l.newToken(LParen, l.ch)
	case ')':
		tok = l.newToken(RParen, l.ch)
	case '{':
		tok = l.newToken(LBrace, l.ch)
	case '}':
		tok = l.newToken(RBrace, l.ch)
	case '[':
		tok = l.newToken(LBracket, l.ch)
	case ']':
		tok = l.newToken(RBracket, l.ch)
	case '"':
		tok.Type = String
		tok.Literal = l.readString()
		return tok
	case 0:
		tok.Type = EOF
		tok.Literal = ""
	default:
		if isLetter(l.ch) {
			tok.Literal = l.readIdentifier()
			tok.Type = LookupIdent(tok.Literal)
			return tok
		}
		if isDigit(l.ch) {
			tok.Type = Int
			tok.Literal = l.readNumber()
			return tok
		}
		l.errors = append(l.errors, fmt.Sprintf("illegal character %q at line %d, column %d", l.ch, l.line, l.column))
		tok = l.newToken(Illegal, l.ch)
	}

	l.readChar()
	return tok
}

func (l *Lexer) newToken(tokenType TokenType, ch rune) Token {
	return Token{
		Type:    tokenType,
		Literal: string(ch),
		Line:    l.line,
		Column:  l.column,
	}
}

func (l *Lexer) readChar() {
	if l.readPosition >= len(l.input) {
		l.ch = 0
		l.position = l.readPosition
		return
	}

	r, size := utf8.DecodeRuneInString(l.input[l.readPosition:])
	l.ch = r
	l.position = l.readPosition
	l.readPosition += size

	if r == '\n' {
		l.line++
		l.column = 0
	} else {
		l.column++
	}
}

func (l *Lexer) peekChar() rune {
	if l.readPosition >= len(l.input) {
		return 0
	}
	r, _ := utf8.DecodeRuneInString(l.input[l.readPosition:])
	return r
}

func (l *Lexer) skipWhitespaceAndComments() {
	for {
		for l.ch == ' ' || l.ch == '\t' || l.ch == '\n' || l.ch == '\r' {
			l.readChar()
		}
		if l.ch == '/' && l.peekChar() == '/' {
			for l.ch != '\n' && l.ch != 0 {
				l.readChar()
			}
			continue
		}
		if l.ch == '/' && l.peekChar() == '*' {
			l.readChar()
			l.readChar()
			for {
				if l.ch == 0 {
					l.errors = append(l.errors, "unterminated block comment")
					return
				}
				if l.ch == '*' && l.peekChar() == '/' {
					l.readChar()
					l.readChar()
					break
				}
				l.readChar()
			}
			continue
		}
		return
	}
}

func (l *Lexer) readIdentifier() string {
	start := l.position
	for isLetter(l.ch) || isDigit(l.ch) {
		l.readChar()
	}
	return l.input[start:l.position]
}

func (l *Lexer) readNumber() string {
	start := l.position
	for isDigit(l.ch) {
		l.readChar()
	}
	return l.input[start:l.position]
}

func (l *Lexer) readString() string {
	l.readChar() // consume opening quote
	start := l.position
	for l.ch != '"' && l.ch != 0 {
		if l.ch == '\\' {
			l.readChar()
			if l.ch == 0 {
				break
			}
		}
		l.readChar()
	}

	if l.ch == 0 {
		l.errors = append(l.errors, fmt.Sprintf("unterminated string starting at line %d, column %d", l.line, l.column))
		return unescape(l.input[start:l.position])
	}

	literal := unescape(l.input[start:l.position])
	l.readChar() // consume closing quote
	return literal
}

func unescape(s string) string {
	var b []rune
	for i := 0; i < len(s); {
		r, size := utf8.DecodeRuneInString(s[i:])
		if r == '\\' && i+size < len(s) {
			next, nextSize := utf8.DecodeRuneInString(s[i+size:])
			switch next {
			case 'n':
				b = append(b, '\n')
			case 't':
				b = append(b, '\t')
			case '\\':
				b = append(b, '\\')
			case '"':
				b = append(b, '"')
			default:
				b = append(b, next)
			}
			i += size + nextSize
			continue
		}
		b = append(b, r)
		i += size
	}
	return string(b)
}

func isLetter(ch rune) bool {
	return unicode.IsLetter(ch) || ch == '_'
}

func isDigit(ch rune) bool {
	return '0' <= ch && ch <= '9'
}

File: vscode-nexus/src/language/lexer.ts
export type TokenType =
  | "ILLEGAL"
  | "EOF"
  | "IDENT"
  | "INT"
  | "STRING"
  | "="
  | "+"
  | "-"
  | "*"
  | "/"
  | "%"
  | "!"
  | "<"
  | ">"
  | "=="
  | "!="
  | "<="
  | ">="
  | ","
  | ";"
  | ":"
  | "::"
  | "=>"
  | "->"
  | "."
  | "|>"
  | "&"
  | "&&"
  | "|"
  | "||"
  | "("
  | ")"
  | "{"
  | "}"
  | "["
  | "]"
  | "LET"
  | "FN"
  | "RETURN"
  | "IF"
  | "ELSE"
  | "TRUE"
  | "FALSE"
  | "NULL"
  | "IMPORT"
  | "ENUM"
  | "MATCH"
  | "ASYNC"
  | "AWAIT"
  | "SPAWN"
  | "CHAN"
  | "MACRO"
  | "RULES"
  | "MUT"
  | "MOVE"
  | "EXTERN"
  | "FROM"
  | "TYPE"
  | "REF"
  | "EFFECT"
  | "PERFORM"
  | "HANDLE"
  | "WITH"
  | "RESUME"
  | "STRUCT"
  | "REGION"
  | "REFLECT"
  | "ALLOC"
  | "DERIVE"
  | "AND"
  | "OR"
  | "NOT"
  | "THEN"
  | "DO"
  | "END"
  | "FUN"
  | "VAR"
  | "WHEN"
  | "IS"
  | "IN"
  | "FOR"
  | "WHILE"
  | "BREAK"
  | "CONTINUE"
  | "PUB"
  | "USE"
  | "HASH";

export interface Token {
  type: TokenType;
  literal: string;
  line: number;
  column: number;
}

const KEYWORDS: Readonly<Record<string, TokenType>> = {
  let: "LET",
  fn: "FN",
  return: "RETURN",
  if: "IF",
  else: "ELSE",
  true: "TRUE",
  false: "FALSE",
  null: "NULL",
  import: "IMPORT",
  enum: "ENUM",
  match: "MATCH",
  async: "ASYNC",
  await: "AWAIT",
  spawn: "SPAWN",
  chan: "CHAN",
  macro: "MACRO",
  rules: "RULES",
  mut: "MUT",
  move: "MOVE",
  extern: "EXTERN",
  from: "FROM",
  type: "TYPE",
  ref: "REF",
  effect: "EFFECT",
  perform: "PERFORM",
  handle: "HANDLE",
  with: "WITH",
  resume: "RESUME",
  struct: "STRUCT",
  region: "REGION",
  reflect: "REFLECT",
  alloc: "ALLOC",
  derive: "DERIVE",
  and: "AND",
  or: "OR",
  not: "NOT",
  then: "THEN",
  do: "DO",
  end: "END",
  fun: "FUN",
  var: "VAR",
  when: "WHEN",
  is: "IS",
  in: "IN",
  for: "FOR",
  while: "WHILE",
  break: "BREAK",
  continue: "CONTINUE",
  pub: "PUB",
  use: "USE",
};

export function lookupIdent(ident: string): TokenType {
  return KEYWORDS[ident] ?? "IDENT";
}

/**
 * Lexer tokenizes Nexus (`.nex`) source into a stream of tokens.
 */
export class Lexer {
  private readonly input: string;
  private position = 0;
  private readPosition = 0;
  private ch = "";
  private line = 1;
  private column = 0;
  private readonly errors: string[] = [];

  constructor(input: string) {
    this.input = input;
    this.readChar();
  }

  getErrors(): readonly string[] {
    return this.errors;
  }

  /** Capture lexer cursor for speculative lookahead. */
  checkpoint(): {
    position: number;
    readPosition: number;
    ch: string;
    line: number;
    column: number;
  } {
    return {
      position: this.position,
      readPosition: this.readPosition,
      ch: this.ch,
      line: this.line,
      column: this.column,
    };
  }

  restore(state: {
    position: number;
    readPosition: number;
    ch: string;
    line: number;
    column: number;
  }): void {
    this.position = state.position;
    this.readPosition = state.readPosition;
    this.ch = state.ch;
    this.line = state.line;
    this.column = state.column;
  }

  nextToken(): Token {
    this.skipWhitespaceAndComments();

    const line = this.line;
    const column = this.column;
    let token: Token;

    switch (this.ch) {
      case "=":
        if (this.peekChar() === "=") {
          const ch = this.ch;
          this.readChar();
          token = { type: "==", literal: ch + this.ch, line, column };
        } else if (this.peekChar() === ">") {
          const ch = this.ch;
          this.readChar();
          token = { type: "=>", literal: ch + this.ch, line, column };
        } else {
          token = this.newToken("=", this.ch, line, column);
        }
        break;
      case "+":
        token = this.newToken("+", this.ch, line, column);
        break;
      case "-":
        if (this.peekChar() === ">") {
          const ch = this.ch;
          this.readChar();
          token = { type: "->", literal: ch + this.ch, line, column };
        } else {
          token = this.newToken("-", this.ch, line, column);
        }
        break;
      case "*":
        token = this.newToken("*", this.ch, line, column);
        break;
      case "/":
        token = this.newToken("/", this.ch, line, column);
        break;
      case "!":
        if (this.peekChar() === "=") {
          const ch = this.ch;
          this.readChar();
          token = { type: "!=", literal: ch + this.ch, line, column };
        } else {
          token = this.newToken("!", this.ch, line, column);
        }
        break;
      case "<":
        if (this.peekChar() === "=") {
          const ch = this.ch;
          this.readChar();
          token = { type: "<=", literal: ch + this.ch, line, column };
        } else {
          token = this.newToken("<", this.ch, line, column);
        }
        break;
      case ">":
        if (this.peekChar() === "=") {
          const ch = this.ch;
          this.readChar();
          token = { type: ">=", literal: ch + this.ch, line, column };
        } else {
          token = this.newToken(">", this.ch, line, column);
        }
        break;
      case "#":
        token = this.newToken("HASH", this.ch, line, column);
        break;
      case "&":
        if (this.peekChar() === "&") {
          const ch = this.ch;
          this.readChar();
          token = { type: "&&", literal: ch + this.ch, line, column };
        } else {
          token = this.newToken("&", this.ch, line, column);
        }
        break;
      case "|":
        if (this.peekChar() === ">") {
          const ch = this.ch;
          this.readChar();
          token = { type: "|>", literal: ch + this.ch, line, column };
        } else if (this.peekChar() === "|") {
          const ch = this.ch;
          this.readChar();
          token = { type: "||", literal: ch + this.ch, line, column };
        } else {
          token = this.newToken("|", this.ch, line, column);
        }
        break;
      case ".":
        token = this.newToken(".", this.ch, line, column);
        break;
      case "%":
        token = this.newToken("%", this.ch, line, column);
        break;
      case ":":
        if (this.peekChar() === ":") {
          const ch = this.ch;
          this.readChar();
          token = { type: "::", literal: ch + this.ch, line, column };
        } else {
          token = this.newToken(":", this.ch, line, column);
        }
        break;
      case ",":
        token = this.newToken(",", this.ch, line, column);
        break;
      case ";":
        token = this.newToken(";", this.ch, line, column);
        break;
      case "(":
        token = this.newToken("(", this.ch, line, column);
        break;
      case ")":
        token = this.newToken(")", this.ch, line, column);
        break;
      case "{":
        token = this.newToken("{", this.ch, line, column);
        break;
      case "}":
        token = this.newToken("}", this.ch, line, column);
        break;
      case "[":
        token = this.newToken("[", this.ch, line, column);
        break;
      case "]":
        token = this.newToken("]", this.ch, line, column);
        break;
      case '"':
        return {
          type: "STRING",
          literal: this.readString(line, column),
          line,
          column,
        };
      case "":
        return { type: "EOF", literal: "", line, column };
      default:
        if (isLetter(this.ch)) {
          const literal = this.readIdentifier();
          return {
            type: lookupIdent(literal),
            literal,
            line,
            column,
          };
        }
        if (isDigit(this.ch)) {
          return {
            type: "INT",
            literal: this.readNumber(),
            line,
            column,
          };
        }
        this.errors.push(
          `illegal character ${JSON.stringify(this.ch)} at line ${line}, column ${column}`,
        );
        token = this.newToken("ILLEGAL", this.ch, line, column);
        break;
    }

    this.readChar();
    return token;
  }

  /** Tokenize the entire input into an array (including trailing EOF). */
  tokenize(): Token[] {
    const tokens: Token[] = [];
    for (;;) {
      const tok = this.nextToken();
      tokens.push(tok);
      if (tok.type === "EOF") {
        break;
      }
    }
    return tokens;
  }

  private newToken(
    type: TokenType,
    ch: string,
    line: number,
    column: number,
  ): Token {
    return { type, literal: ch, line, column };
  }

  private readChar(): void {
    if (this.readPosition >= this.input.length) {
      this.ch = "";
      this.position = this.readPosition;
      return;
    }

    const ch = this.input[this.readPosition]!;
    this.ch = ch;
    this.position = this.readPosition;
    this.readPosition += 1;

    if (ch === "\n") {
      this.line += 1;
      this.column = 0;
    } else {
      this.column += 1;
    }
  }

  private peekChar(): string {
    if (this.readPosition >= this.input.length) {
      return "";
    }
    return this.input[this.readPosition]!;
  }

  private skipWhitespaceAndComments(): void {
    for (;;) {
      while (
        this.ch === " " ||
        this.ch === "\t" ||
        this.ch === "\n" ||
        this.ch === "\r"
      ) {
        this.readChar();
      }

      const ch = this.ch;
      const peek = this.peekChar();

      if (ch === "/" && peek === "/") {
        this.readChar();
        for (;;) {
          const c = this.currentChar();
          if (c === "\n" || c === "") {
            break;
          }
          this.readChar();
        }
        continue;
      }

      if (ch === "/" && peek === "*") {
        this.readChar(); // /
        this.readChar(); // *
        for (;;) {
          const c = this.currentChar();
          if (c === "") {
            break;
          }
          if (c === "*" && this.peekChar() === "/") {
            this.readChar();
            this.readChar();
            break;
          }
          this.readChar();
        }
        continue;
      }

      break;
    }
  }

  /** Returns the current character without control-flow narrowing side effects. */
  private currentChar(): string {
    return this.ch;
  }

  private readIdentifier(): string {
    const start = this.position;
    while (isLetter(this.ch) || isDigit(this.ch)) {
      this.readChar();
    }
    return this.input.slice(start, this.position);
  }

  private readNumber(): string {
    const start = this.position;
    while (isDigit(this.ch)) {
      this.readChar();
    }
    return this.input.slice(start, this.position);
  }

  private readString(startLine: number, startColumn: number): string {
    this.readChar(); // opening quote
    let value = "";

    for (;;) {
      const c = this.ch;
      if (c === '"' || c === "") {
        break;
      }

      if (c === "\\") {
        this.readChar();
        const escaped = this.ch;
        switch (escaped) {
          case "n":
            value += "\n";
            break;
          case "t":
            value += "\t";
            break;
          case "r":
            value += "\r";
            break;
          case "\\":
            value += "\\";
            break;
          case '"':
            value += '"';
            break;
          case "":
            break;
          default:
            value += escaped;
            break;
        }
        if (this.ch !== "") {
          this.readChar();
        }
        continue;
      }

      value += c;
      this.readChar();
    }

    if (this.ch === "") {
      this.errors.push(
        `unterminated string starting at line ${startLine}, column ${startColumn}`,
      );
      return value;
    }

    this.readChar(); // closing quote
    return value;
  }
}

function isLetter(ch: string): boolean {
  return (
    (ch >= "a" && ch <= "z") ||
    (ch >= "A" && ch <= "Z") ||
    ch === "_" ||
    ch === "$"
  );
}

function isDigit(ch: string): boolean {
  return ch >= "0" && ch <= "9";
}
